matrix/basic_matrix.py

```diff
@@ -189,11 +201,108 @@ def diagonal(self):
             result.append(self.ls_entries[i][i])
         return result
 
+    def is_zero_matrix(self):
+        return all(x == 0 for x in chain(*self.ls_entries))
+
+    @staticmethod
+    def reduced_row_echelon_form(A):
+        """
+        Given Matrix A, return the reduced row echelon form
+
+        Args:
+            A: <Matrix> to put into row echelon form
+        Returns:
+            <Matrix> in reduced row echelon form
+        """
+        A = Matrix.row_echelon_form(A)
+        len_col, len_row = len(A.ls_entries[0]), len(A.ls_entries)
+        loop_from = len_col if len_col == len_row else len_col - 1
+
+        for i in reversed(range(1, loop_from)):
+            reduction_fact = Matrix([[x * z[0] for x in A[i]] for z in zip(A[:i, i])])
+            A[:i] = Matrix(A[:i]).__add__(reduction_fact.__neg__())
```

Comment:
Not sure if this makes sense, but should this be using matrix factory instead of Matrix?

Reply:
nevermind, this is fine.